drivers/gpio/gpio_infineon.c

```diff
@@ -158,26 +182,51 @@ static uint32_t gpio_cat1_get_pending_int(const struct device *dev)
 	const struct gpio_cat1_config *const cfg = dev->config;
 	GPIO_PRT_Type *const base = cfg->regs;
 
+#if defined(CONFIG_SOC_FAMILY_INFINEON_PSOC4)
+	return base->INTR & gpio_cat1_valid_mask(cfg->ngpios);
+#else
 	return GPIO_PRT_INTR_MASKED(base);
+#endif
 }
 
 #if (!(CONFIG_SOC_FAMILY_INFINEON_CAT1C && CONFIG_CPU_CORTEX_M0PLUS))
 static void gpio_isr_handler(const struct device *dev)
 {
 	const struct gpio_cat1_config *const cfg = dev->config;
+	struct gpio_cat1_data *const data = dev->data;
 	GPIO_PRT_Type *const base = cfg->regs;
-	uint32_t pins = GPIO_PRT_INTR_MASKED(base);
+	uint32_t pending;
 
-	for (uint8_t i = 0; i < CY_GPIO_PINS_MAX; i++) {
-		Cy_GPIO_ClearInterrupt(base, i);
+#if defined(CONFIG_SOC_FAMILY_INFINEON_PSOC4)
+	uint32_t intr_status = base->INTR;
+	uint32_t intr_mask = 0U;
+
+	for (uint8_t i = 0; i < cfg->ngpios; i++) {
+		uint32_t intr_cfg = (base->INTR_CFG >> (i * 2U)) & 0x3U;
+
+		if (intr_cfg != CY_GPIO_INTR_DISABLE) {
+			intr_mask |= BIT(i);
+		}
 	}
 
-	if (dev) {
-		gpio_fire_callbacks(&((struct gpio_cat1_data *const)(dev)->data)->callbacks, dev,
-				    pins);
+	pending = intr_status & intr_mask;
```

Comment:
Shouldn't we be able to mask against the number of pins on the port without inspecting the interrupt configure bits? I noticed this too with #98564 (which this likely will conflict with).